# Exploring eBay Car Sales Data

Darren Ho

## Introduction

In this project, we will work with a dataset of used cars from eBay *Kleinanzeigen*, a classified section of the German eBay website. 

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset is not available on Kaggle anymore, but can be find [here](https://data.world/data-society/used-cars-data)

There have a been a few modifications made to the original dataset:

- Sampled 50,000 data points from the full dataset, to ensure code runs quickly in hosted environment
- Dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary can be found in the link [above](https://data.world/data-society/used-cars-data).

The aim of this specific project is to clean the data and analyze the included used car listings. Let's start by importing the libraries we need and reading the dataset into pandas. 

In [139]:
#importing libraries

import numpy as np
import pandas as pd

In [140]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [141]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [142]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There are 20 columns, 15 of which are of the object type and the remaining 5 are of the int type. Most of the columns are strings. We also see that some columns have null values, but none have more than ~20% null values.  

## Cleaning Column Names

The column names in the dataset use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we cannot just replace spaces with underscores. 

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [143]:
# Print an array of the existing column names 

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We will copy the array and make the following edits to column names:

- `yearOfRegistration` to `registration_year`
- `monthOfRegistration` to `registration_month`
- `notRepairedDamage` to `unrepaired_damage`
- `dateCreated` to `ad_created`
- The rest of the column names from camelcase to snakecase.

In [144]:
# Assigning modified column names back to the autos.columns attribute

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postal_code',
       'last_seen']

In [145]:
autos.head() # Looking at the current state of the autos dataframe

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning

Now we can do some basic data exploration to determine what other cleaning tasks may need to be done. Initially, we'll look for:

- Text columns where (all or almost) values are the same
- Examples of numberic data stored as text which can be cleaned and converted 



In [146]:
# Looking at descriptive stats for all columns 

autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Taking a closer look, we see that the `seller` and `offer_type` columns have almost all the same values, so we can probably drop these variables. 

Taking a closer look at the `registration` variables, we see that there are some weird values. In the case of `registration_year`, we see a minimum year of 1000 and a maximum year of 9999. As for `registration_month`, we see that a month of 0 is the most frequent value in the variable.

We also notice that the `num_of_pictures` column is completetly full of 0's, so we could probably drop this variable as well. 

We found that the `price` and `odometer` columns are numeric values stored as text. So for now, let's clean these columns by doing the following: 

- Remove any non-numeric characters
- Convert the column to a numeric dtype
- Rename the odometer column

In [147]:
# First, lets drop the mentioned variables

autos = autos.drop(['seller','offer_type','num_of_pictures'], axis = 1)

In [148]:
# Removing non-numeric characters and converting to numeric dtype

autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','').astype(int)
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','').astype(int)

In [149]:
# Renaming odometer column

autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

autos['price'].head(5)  # checking

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [150]:
autos['odometer_km'].head(5)  # checking

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Exploring the Odometer and Price Columns

Let's continue exploring the data, specifically look for data that does not look right. We'll start by analyzing the `odometer_km` and `price` columns. Here are is what we will do:

- Analyze the columns using min and max values & look for any outliers that we may want to remove

In [165]:
#odometer

print("Unique Values:",autos['odometer_km'].unique().shape) # see how many unique values
print('\n')
print(autos['odometer_km'].describe())       # view min/max/median/mean
print('\n')
print(autos['odometer_km'].value_counts().head(13).sort_index(ascending = True)) 
print('\n')
print(autos['odometer_km'].value_counts().head(13).sort_index(ascending = False)) 

Unique Values: (13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64


Using the techniques above, we find that there are only 13 unique values in the `odometer_km` column. This probably means that all the odometers were recorded as a rounded number. We find the 5-number summary for the variable as well. We then return a series, one with `ascending = True` and another variation with `ascending = False` to view the highest and lowest values with their counts. There does not appear to be any values that look unrealistically high or low that we might want to remove.  

In [166]:
# Price

print("Unique Values:",autos['price'].unique().shape) # see how many unique values
print('\n')
print(autos['price'].describe())       # view min/max/median/mean
print('\n')
print(autos['price'].value_counts().sort_index(ascending = True).head(20))
print('\n')
print(autos['price'].value_counts().sort_index(ascending = False).head(20))

Unique Values: (2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


Following up with the `price` column, we see that this variable has 2357 unique values. Similar to the `odometer_km` column, the values recorded in this column are also rounded. Unlike `odometer_km`, we actually have outliers that we might want to remove. Running `value_counts` and then sorting by highest/lowest values, we see that there were 1421 ad prices of \\$0, and even several data points over \\$1 million. 

Let's remove the outliers. We'll get rid of the eBay ads that have listing prices of \\$0 and anything above \\$350,000 as that seems to be the point before prices skyrocket astronimcally. 

In [169]:
# removing outliers

autos = autos[autos["price"].between(1,350000)]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

We removed about 1,435 data points that we considered outliers. The mean price of the remaining values is now approximately \\$5,889, compared to the original mean price of \\$9,840. Standard deviation has also decreased from \\$481,104 to a more realistic \\$9060.     

## Exploring the Date Columns

Let's now move on to the date columns and understand the date range the data covers. 

There are 5 columns that should represent date values. Some were created by the crawler, some came from the webiste itself. 

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

`date_crawled`, `last_seen`, `ad_created` columns are all identified as a string values by pandas, which means we need to convert the data into a numerical representation so we can understand it quantitatively. Let's first understand how the values in the three string columns are formatted, as they represent full timestamp values

In [172]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We notice that the first 10 characters represent the day. To understand the date range, we can extract just the date values.

In [183]:
# date_crawled 
# calc the distribution of values as percentages

(autos['date_crawled'].str[:10]               # selecting first 10 characters in each column
                     .value_counts(normalize=True, dropna=False)     # include missing values in distribution and to use percentages instead of counts 
                     .mul(100)
                     .round(2)
                     .sort_index())      # to rank by date in ascending order (earliest to latest)

2016-03-05    2.53
2016-03-06    1.40
2016-03-07    3.60
2016-03-08    3.33
2016-03-09    3.31
2016-03-10    3.22
2016-03-11    3.26
2016-03-12    3.69
2016-03-13    1.57
2016-03-14    3.65
2016-03-15    3.43
2016-03-16    2.96
2016-03-17    3.16
2016-03-18    1.29
2016-03-19    3.48
2016-03-20    3.79
2016-03-21    3.74
2016-03-22    3.30
2016-03-23    3.22
2016-03-24    2.93
2016-03-25    3.16
2016-03-26    3.22
2016-03-27    3.11
2016-03-28    3.49
2016-03-29    3.41
2016-03-30    3.37
2016-03-31    3.18
2016-04-01    3.37
2016-04-02    3.55
2016-04-03    3.86
2016-04-04    3.65
2016-04-05    1.31
2016-04-06    0.32
2016-04-07    0.14
Name: date_crawled, dtype: float64

eBay appears to have been crawled daily from the beginning of March 2016 into the first week of April 2016. The highest percentage of 3.86% can be found on 2016-04-03.

In [184]:
# ad_created 
# calc the distribution of values as percentages

(autos['ad_created'].str[:10]               # selecting first 10 characters in each column
                     .value_counts(normalize=True, dropna=False)     # include missing values in distribution and to use percentages instead of counts 
                     .mul(100)
                     .round(2)                                               
                     .sort_index())      # to rank by date in ascending order (earliest to latest)

2015-06-11    0.00
2015-08-10    0.00
2015-09-09    0.00
2015-11-10    0.00
2015-12-05    0.00
              ... 
2016-04-03    3.89
2016-04-04    3.69
2016-04-05    1.18
2016-04-06    0.33
2016-04-07    0.13
Name: ad_created, Length: 76, dtype: float64

There are 76 unique dates for which the ad on eBay was created. The earliest date in the `ad_created` column is 2015-06-11, while the latest date is 2016-04-07 

In [185]:
# last_seen 
# calc the distribution of values as percentages

(autos['last_seen'].str[:10]               # selecting first 10 characters in each column
                     .value_counts(normalize=True, dropna=False)     # include missing values in distribution and to use percentages instead of counts 
                     .mul(100)
                     .round(2)                                              
                     .sort_index())      # to rank by date in ascending order (earliest to latest)

2016-03-05     0.11
2016-03-06     0.43
2016-03-07     0.54
2016-03-08     0.74
2016-03-09     0.96
2016-03-10     1.07
2016-03-11     1.24
2016-03-12     2.38
2016-03-13     0.89
2016-03-14     1.26
2016-03-15     1.59
2016-03-16     1.65
2016-03-17     2.81
2016-03-18     0.74
2016-03-19     1.58
2016-03-20     2.07
2016-03-21     2.06
2016-03-22     2.14
2016-03-23     1.85
2016-03-24     1.98
2016-03-25     1.92
2016-03-26     1.68
2016-03-27     1.56
2016-03-28     2.09
2016-03-29     2.23
2016-03-30     2.48
2016-03-31     2.38
2016-04-01     2.28
2016-04-02     2.49
2016-04-03     2.52
2016-04-04     2.45
2016-04-05    12.48
2016-04-06    22.18
2016-04-07    13.19
Name: last_seen, dtype: float64

The `last_seen` column represents the last time a crawler saw a specific ad. With that being said, we notice a large portion of the distribution happens in the last 3 consecutive days. 

In [186]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As noted before, registration is unusual in that the maximum year registration year is 9999 while the minimum year is 1000. Let's fix this. 

## Dealing with Incorrect Registration Year Data

Because a car cannot be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listing with cars that fall outside the 1900-2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [206]:
autos = autos[autos['registration_year'].between(1900,2016)]

autos['registration_year'].value_counts(normalize = True).mul(100).round(2).head(10) 

# normalize = true helps remove the values outside those upper/lower bounds and calc distrib of
# remaining values

2000    6.76
2005    6.29
1999    6.21
2004    5.79
2003    5.78
2006    5.72
2001    5.65
2002    5.33
1998    5.06
2007    4.88
Name: registration_year, dtype: float64

Removing years before 1900 and year after 2016 from the column, we see that a large majority of cars were first registered just recently. The 20 years leading up to 2016 covers over 50% of cars that were registered. 

## Exploring Price by Brand 

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the `brand` column. 

In [212]:
print(autos['brand'].describe())
print('\n')
print(autos['brand'].unique())

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object


['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


In [215]:
autos['brand'].value_counts(normalize = True).mul(100).round(2)

volkswagen        21.13
bmw               11.00
opel              10.76
mercedes_benz      9.65
audi               8.66
ford               6.99
renault            4.71
peugeot            2.98
fiat               2.56
seat               1.83
skoda              1.64
nissan             1.53
mazda              1.52
smart              1.42
citroen            1.40
toyota             1.27
hyundai            1.00
sonstige_autos     0.98
volvo              0.91
mini               0.88
mitsubishi         0.82
honda              0.78
kia                0.71
alfa_romeo         0.66
porsche            0.61
suzuki             0.59
chevrolet          0.57
chrysler           0.35
dacia              0.26
daihatsu           0.25
jeep               0.23
subaru             0.21
land_rover         0.21
saab               0.16
jaguar             0.16
daewoo             0.15
trabant            0.14
rover              0.13
lancia             0.11
lada               0.06
Name: brand, dtype: float64

Exploring the unique values in the `brand` column, we see that there are 40 unique brands with Volkswagen being the most frequently sold. When we list the brands by distribution, from highest to lowest, we see that the top 5 most sold brands are all German.

We have a lot of brands that are not significant in terms of distribution share, so we'll limit our analysis to the the brands that exceed 5%.

In [222]:
brand_list = autos['brand'].value_counts(normalize = True).mul(100).round(2).head(6).index
brand_list

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [223]:
# aggregating data

mean_prices = {}

for b in brand_list:
    brand_only = autos[autos["brand"] == b]
    mean_price = brand_only["price"].mean()
    mean_prices[b] = int(mean_price)
    
mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Of the top 6 brands, the most expensive brands consist of: Audi, Mercedes Benz, and BMW. From there, there is a gap between the 3 most expensive brands to the remaining brands. The cheapest brand in the top 6 is Opel, followed by Ford and then Volkswagen. 

## Exploring Mileage by Brand

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.

In [227]:
bmp_series = pd.Series(mean_prices)
pd.DataFrame(bmp_series, columns=['mean_price'])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [229]:
mean_mileage_by_brand = {}

for b in brand_list:
    brand_only = autos[autos["brand"] == b]
    mean_mileage = brand_only["odometer_km"].mean()
    mean_mileage_by_brand[b] = int(mean_mileage)
    
mean_mileage_by_brand

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [234]:
mm = pd.Series(mean_mileage_by_brand).sort_values(ascending=False)
mp = pd.Series(mean_prices).sort_values(ascending=False)

brand_info = pd.DataFrame(mm,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [235]:
brand_info["mean_price"] = mp
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

## Conclusion

Of the top 6 brands, BMWs had the highest mean milage at 132,572 km whereas Fords had the lowest mean mileage at 124,266 km. In terms of mean price, BMWs ranks 3rd while Fords rank 5th (expensive to cheap). 

## Next Steps

In this guided project, 